move has_torch_function to C++, and make a special case object_has_torch_function

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48966 Treat has_torch_function and object_has_torch_function as static False when scripting
* **#48965 move has_torch_function to C++, and make a special case object_has_torch_function**
* #48964 clean up imports for tensor.py
* #48963 treat Parameter the same way as Tensor

This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590732](https://our.internmc.facebook.com/intern/diff/D25590732)

Commit:
move has_torch_function to C++, and make a special case object_has_torch_function

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

[ghstack-poisoned]

robieta commented:
Ahhh. `PySequence_Fast` creates a new ref, it doesn't borrow.

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

[ghstack-poisoned]

robieta commented:
TODO: The logic in this PR should subsume the `__torch_function__` logic in `python_arg_parser`

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

[ghstack-poisoned]

hameerabbasi commented on torch/_C/__init__.pyi.in:
```diff
@@ -494,6 +494,8 @@ def _get_qengine() -> _int: ...  # THPModule_qEngine
 def _set_qengine(qegine: _int) -> None: ...  # THPModule_setQEngine
 def _supported_qengines() -> List[_int]: ...  # THPModule_supportedQEngines
 def _is_xnnpack_enabled() -> _bool: ...  # THPModule_isEnabledXNNPACK
+def _has_torch_function(Any) -> _bool: ...  # THPModule_has_torch_function
```
Does this work?
```suggestion
def _has_torch_function(Iterable[Any]) -> _bool: ...  # THPModule_has_torch_function
```

robieta replied:
Good point.

hameerabbasi commented on torch/csrc/autograd/python_variable.h:
```diff
@@ -24,14 +24,17 @@ THP_API PyObject *ParameterClass;
+static inline bool THPVariable_CheckExact(PyTypeObject* tp) {
   return (
-    obj_py_type == (PyTypeObject*)THPVariableClass ||
-    obj_py_type == (PyTypeObject*)ParameterClass
+    tp == (PyTypeObject*)THPVariableClass ||
+    tp == (PyTypeObject*)ParameterClass
   );
 }
 
+inline bool THPVariable_CheckExact(PyObject *obj) {
+  return THPVariable_CheckExact(Py_TYPE(obj));
+}
+
```
I wouldn't do this. Sometimes, `PyTypeObject*` can be casted as a `PyObject*`, and call the wrong overload here (since they C-inherit from one another). In such a case, you may end up checking for the equivalent of `has_torch_function(type(type(obj)))`.

robieta replied:
Good catch!!! Serves me right for trying to be clever with overloads.

robieta replied:
Changed to be two different functions rather than an overload.

hameerabbasi commented on torch/csrc/utils/disable_torch_function.cpp:
```diff
@@ -125,3 +126,102 @@ PyObject* THPModule_disable_torch_function(PyObject *self, PyObject *a) {
+// Makes sure that we don't check for __torch_function__ on basic Python types
+static bool is_basic_python_type(PyTypeObject *tp)
+{
+  return (
```
A tangent, but this is a place I see `constexpr unordered_map` coming in handy in the future.

robieta replied:
I actually tried `ska::flat_hash_set` but it was slower than just doing the checks directly.

hameerabbasi commented on torch/csrc/utils/disable_torch_function.cpp:
```diff
@@ -125,3 +126,102 @@ PyObject* THPModule_disable_torch_function(PyObject *self, PyObject *a) {
+    Py_RETURN_FALSE;
+  }
+
+  PyObject* args(PySequence_Fast(arg, "expected a sequence"));
+  auto result = sequence_has_torch_function(args);
+  Py_DECREF(args);
```
Perhaps use `py::object` for auto refcounting?

robieta replied:
Done.

hameerabbasi commented on torch/csrc/utils/python_strings.h:
```diff
@@ -54,3 +55,49 @@ inline bool THPUtils_isInterned(PyObject* obj) {
+
+static py::object PyObject_FastGetAttrString(PyObject *obj, char *name)
+{
+    PyTypeObject *tp = Py_TYPE(obj);
```
We're possibly leaking `tp` here.

robieta replied:
How so? (This is the same as what numpy does, FWIW.)

hameerabbasi replied:
Never mind, I was thinking of [`PyObject_Type`](https://docs.python.org/3/c-api/object.html?highlight=pyobject_type#c.PyObject_Type) instead of [`Py_TYPE`](https://docs.python.org/3/c-api/structures.html?highlight=py_type#c.Py_TYPE).

hameerabbasi commented on torch/csrc/utils/python_strings.h:
```diff
@@ -54,3 +55,49 @@ inline bool THPUtils_isInterned(PyObject* obj) {
+    }
+    /* Attribute referenced by (PyObject *)name */
+    else if (tp->tp_getattro != NULL) {
+        PyObject *w = THPUtils_internString(name);
+        if (w == NULL) {
+          return py::object();
+        }
+        res = (*tp->tp_getattro)(obj, w);
+        Py_DECREF(w);
```
Again, `py::object` would be really nice. Tbf, it isn't really your code, but fixups can't hurt. 😄

robieta replied:
Done.

hameerabbasi reviewed (changes requested):
A couple of comments, most of the logic looks good!

ezyang commented:
There is code movement in this PR, could you please import it into Phabricator (which shows when code fragments are moved) or please comment on all locations where code was moved without changes.

ezyang commented on torch/functional.py:
```diff
@@ -965,7 +965,7 @@ def tensordot(a, b, dims=2, out=None):
     """
     if not torch.jit.is_scripting():
-        if (type(a) is not Tensor or type(b) is not Tensor) and has_torch_function((a, b)):
+        if has_torch_function((a, b)):
```
This seems like a pessimization. Previously, tuple construction only happened if a was not a tensor and b was not a tensor. Now you unconditionally do tuple construction so you can call `has_torch_function`. I'm not too sure how CPython internally implements variadic function calls, but `has_torch_function(a, b)` has the potential to bypass this (although I guess if in the variadic case CPython constructs a tuple anyway, you'll need separate overloads per arity.)

robieta replied:
Surprisingly it's not. (kind of)

```
import re
from torch.utils.benchmark import Timer

for stmt in (
    "type(a) is not Tensor",
    "type(a) is not Tensor or type(b) is not Tensor",
    "object_has_torch_function(a)",
    "object_has_torch_function(a) or object_has_torch_function(b)",
    "has_torch_function((a, b))"
):
    timer = Timer(
        stmt,
        setup="""
        from torch import Tensor
        from torch.overrides import has_torch_function, object_has_torch_function
        a = torch.ones((1,))
        b = torch.ones((1,))
        # end_setup
        """
    )
    # Trim repeated `setup` for readability.
    s = repr(timer.blocked_autorange(min_run_time=2))
    print(re.sub(r"setup:.+# end_setup\n\n", "", s, flags=re.DOTALL), "\n")
```

```
<torch.utils.benchmark.utils.common.Measurement object at 0x7f254c3c5190>
type(a) is not Tensor
  Median: 73.92 ns
  IQR:    3.14 ns (72.39 to 75.52)
  268 measurements, 100000 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7f254c546c50>
type(a) is not Tensor or type(b) is not Tensor
  Median: 157.44 ns
  IQR:    6.42 ns (153.04 to 159.46)
  128 measurements, 100000 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7f254c3c5cd0>
object_has_torch_function(a)
  Median: 31.67 ns
  IQR:    1.40 ns (31.22 to 32.61)
  624 measurements, 100000 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7f2571ba7a50>
object_has_torch_function(a) or object_has_torch_function(b)
  Median: 60.68 ns
  IQR:    4.44 ns (58.10 to 62.55)
  323 measurements, 100000 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7f254c3c5bd0>
has_torch_function((a, b))
  Median: 65.73 ns
  IQR:    4.23 ns (63.92 to 68.16)
  296 measurements, 100000 runs per measurement, 1 thread
```

Even though the previous version wouldn't create a tuple for Tensor args, tuples are not that expensive and doing things in Python is slow. I checked, and while CPython has `METH_O`, there's no `METH_1` for static two arg functions so you'd have to go through `METH_VARARGS` and the tuple (or maybe `list`!!!) anyway. That said, one thing that I could do is to make a private version that statically expects a size two tuple as well to skip some of the unpacking work on the C++ side. (And MyPy will keep us honest on the calling side.) It won't be much, but since this is called everywhere it might be worth it to shave some nanoseconds.

robieta replied:
Ohh, actually `METH_FASTCALL` might be the thing.

robieta replied:
(It's 3.7+ though, so if it's useful it would have to be version guarded.)

ezyang replied:
FASTCALL seems like the right thing, although `object_has_torch_function(a) or object_has_torch_function(b)` doesn't seem too shabby either. So while we're ripping up the floorboards right now, might as well design the API so that it is FASTCALL compatible.

robieta replied:
`FASTCALL` times:

```
<torch.utils.benchmark.utils.common.Measurement object at 0x7f5f511e3a90>
_has_torch_function_varargs(a)
  Median: 40.99 ns
  IQR:    1.53 ns (40.50 to 42.03)
  475 measurements, 100000 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7f5f75c649d0>
_has_torch_function_varargs(a, b)
  Median: 46.04 ns
  IQR:    1.33 ns (45.49 to 46.82)
  433 measurements, 100000 runs per measurement, 1 thread
```

So it's worth it to keep the single arg version, but for multiple it's considerably better to skip the intermediate tuple. And for Python 3.6 we can just use `METH_VARARGS` which isn't the end of the world. Particularly since we can statically use the tuple fast path that has to be checked in `THPModule_has_torch_function`.

ezyang replied:
we're dumping py3.6 soon anyway

ezyang replied:
By the way, I bet we have a lot of existing bindings that use `METH_VARARGS` and can switch to FASTCALL. So we should file a bug about this too.

robieta replied:
https://github.com/pytorch/pytorch/issues/49476

ezyang commented on torch/overrides.py:
```diff
@@ -1182,33 +1183,36 @@ def handle_torch_function(
+      `object_has_torch_function(t)`
+    which skips unnecessary packing and unpacking work.
+    """
+)
```
Yeah, it seems better to name the functions based on arity, makes it more obvious that `object_has_torch_function` is for one-input case (as opposed to being some object-y thing)

robieta replied:
My conundrum is that they're all _technically_ unary functions. That said, if we do also go for the special case size two tuple case then something like `unary_has_torch_function` and `unsafe_binary_has_torch_function` might be clearer. Feel free to suggest better names if any occur to you.

ezyang replied:
`has_torch_function1`, `has_torch_function2`, etc. etc. But I should actually respond to your comments on the other thread. Hang on...

ezyang commented on torch/overrides.py:
```diff
@@ -167,6 +167,7 @@ def get_ignored_functions() -> Set[Callable]:
         torch.nn.functional.upsample_nearest,
         torch.nn.functional.has_torch_function,
+        torch.nn.functional.object_has_torch_function,
         torch.nn.functional.handle_torch_function,
```
??? Are these getting secretly exported when they shouldn't be ???

robieta replied:
I was surprised too, but yeah. Tests fail if I don't.

ezyang replied:
Can you file an issue about this? They shouldn't be exported... you might be able to fix this problem by importing the module rather than the functions; maybe `from blah import foo` implicitly sets up an export.

robieta replied:
https://github.com/pytorch/pytorch/issues/49480

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

[ghstack-poisoned]

hameerabbasi commented on torch/csrc/utils/python_strings.h:
```diff
@@ -54,3 +55,49 @@ inline bool THPUtils_isInterned(PyObject* obj) {
+static py::object PyObject_FastGetAttrString(PyObject *obj, char *name)
+{
+    PyTypeObject *tp = Py_TYPE(obj);
+    PyObject *res = (PyObject *)NULL;
+
+    /* Attribute referenced by (char *)name */
+    if (tp->tp_getattr != NULL) {
+        res = (*tp->tp_getattr)(obj, name);
+        if (res == NULL) {
+          PyErr_Clear();
+        }
+    }
+    /* Attribute referenced by (PyObject *)name */
+    else if (tp->tp_getattro != NULL) {
+        auto w = py::reinterpret_steal<py::object>(
+          THPUtils_internString(name));
+        if (w.ptr() == nullptr) {
+          return py::object();
+        }
+        res = (*tp->tp_getattro)(obj, w.ptr());
+        if (res == NULL) {
+            PyErr_Clear();
+        }
+    }
+    return py::reinterpret_steal<py::object>(res);
```
Nit: Can we change `NULL` to `nullptr`?

robieta replied:
Done. I wasn't sure if I needed to leave NULL for C compat.

hameerabbasi reviewed (commented):
Will approve as soon as CI passes, but this might be fixable right away.

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590732](https://our.internmc.facebook.com/intern/diff/D25590732)

[ghstack-poisoned]

robieta commented:
> There is code movement in this PR, could you please import it into Phabricator (which shows when code fragments are moved) or please comment on all locations where code was moved without changes.

Done.

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590732](https://our.internmc.facebook.com/intern/diff/D25590732)

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590732](https://our.internmc.facebook.com/intern/diff/D25590732)

[ghstack-poisoned]

robieta commented:
> Will approve as soon as CI passes, but this might be fixable right away.

Tests are now passing, modulo some noise on internal tests.

hameerabbasi reviewed (approved):
I've gone through this with as fine toothed a comb as I can, modulo maybe the `METH_FASTCALL` wrapper.

robieta commented:
As a follow up to https://github.com/pytorch/pytorch/pull/48965#discussion_r542184193, I did some experiments with [frozen](https://github.com/serge-sans-paille/frozen) and it is insanely fast. The std maps don't constexpr so I had to use const (constexpr-ability is pretty much the main selling point of frozen) and I just tweaked the benchmarks.

The benchmark was pretty simple; fetch a key from a size four map of ints. (Since PyType will be a pointer, python strings will be an interned pointer, etc so I expect this to be pretty representative.)

```
BM_IntInFzMap                   0.357 ns        0.357 ns   1000000000
BM_IntInStdMap                   31.4 ns         31.4 ns     22841712
BM_IntInStdUnorderedMap          61.7 ns         61.7 ns     11482749
```

Taking a step back, I could see this being really powerful in conjunction with [METH_FASTCALL | METH_KEYWORDS](https://docs.python.org/3/c-api/structures.html#METH_FASTCALL) to really squeeze the python arg parser.

CC @ezyang @ljk53 @bhosmer @ngimel

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590732](https://our.internmc.facebook.com/intern/diff/D25590732)

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590732](https://our.internmc.facebook.com/intern/diff/D25590732)

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/disable_torch_function.cpp:
```diff
@@ -125,3 +126,111 @@ PyObject* THPModule_disable_torch_function(PyObject *self, PyObject *a) {
 }
+
+// Makes sure that we don't check for __torch_function__ on basic Python types
+static bool is_basic_python_type(PyTypeObject *tp)
```
This function looks unchanged but the functions below seem to have been changed

robieta replied:
Indeed, this one was just moved from `torch/csrc/utils/python_arg_parser.h` The functions below were modified to:
a) Bail out faster. (e.g. don't do attr checks on known types)
b) More efficiently handle checking of multiple Python values, which generally means trying to be as lazy as possible with Python containers. (e.g. `PySequence_Fast` does an extra refcount bump and decref because it has no way of knowing that we'll keep args alive until it's done.)

ezyang commented on torch/csrc/utils/disable_torch_function.cpp:
```diff
@@ -125,3 +126,111 @@ PyObject* THPModule_disable_torch_function(PyObject *self, PyObject *a) {
+  return (
+    !THPVariable_CheckTypeExact(tp) &&
+    !is_basic_python_type(tp) &&
+    torch::torch_function_enabled() &&
```
huh, I would have expected `torch::torch_function_enabled` to be the first thing to test. Is it more expensive than I thought?

robieta replied:
The main reason is that I expect `!THPVariable_CheckTypeExact(tp) && !is_basic_python_type(tp)` to be false in most cases (particularly in normal eager use) while `torch::torch_function_enabled()` is normally true, so it was a question of `likely_false && likely_true` instead of the other way around. I _think_ I checked and it saved an instruction or two, although between branch predictors and instruction parallelism I don't know if it actually matters on a real chip. And because `sequence_has_torch_function` calls `has_torch_function`, you could wind up doing the check a whole **two to three** times!!! (Alas, we're waaaaayyyy under what I could hope to A/B with wall time here.) It probably doesn't matter, but this part of the path is so hot that I'm paranoid.

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590732](https://our.internmc.facebook.com/intern/diff/D25590732)

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590732](https://our.internmc.facebook.com/intern/diff/D25590732)

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590732](https://our.internmc.facebook.com/intern/diff/D25590732)

[ghstack-poisoned]

Commit:
Update on "move has_torch_function to C++, and make a special case object_has_torch_function"


This PR pulls `__torch_function__` checking entirely into C++, and adds a special `object_has_torch_function` method for ops which only have one arg as this lets us skip tuple construction and unpacking. We can now also do away with the Python side fast bailout for `Tensor` (e.g. `if any(type(t) is not Tensor for t in tensors) and has_torch_function(tensors)`) because they're actually slower than checking with the Python C API.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590732](https://our.internmc.facebook.com/intern/diff/D25590732)

[ghstack-poisoned]
